### PR TITLE
Implement archive record function

### PR DESCRIPTION
**Pull request by smarlowucf**
And integrate archive into the adapter event loop. Each billing cycle is archived with metering and usage info included. With a default archive length of 6 months.

### CODE REVIEW

**Comment by smarlowucf** [comment]
Example archive structure for one "month" of data:

```
[
  {
        'billing_time': '2024-02-09T18:11:59.527064+00:00',
        'billing_status': {
            'tier_1': {
                'record_id': 'd92c6e6556b14770994f5b64ebe3d339',
                'status': 'succeeded'
            }
        },
        'billed_usage': {
            'tier_1': 10
        },
        'usage_records': [
            {
                'managed_node_count': 9,
                'reporting_time': '2024-01-09T18:11:59.527673+00:00',
                'base_product': 'cpe:/o:suse:product:v1.2.3'
            },
            {
                'managed_node_count': 9,
                'reporting_time': '2024-01-09T18:11:59.529096+00:00',
                'base_product': 'cpe:/o:suse:product:v1.2.3'
            },
            {
                'managed_node_count': 10,
                'reporting_time': '2024-01-09T18:11:59.531424+00:00',
                'base_product': 'cpe:/o:suse:product:v1.2.3'
            }
        ]
    }
]
```

File: csp_billing_adapter/archive.py [hunk 1]
```diff
@@ -45,3 +55,48 @@ def append_metering_records(
...
+        functools.partial(
+            hook.get_metering_archive,
+            config=config,
+        ),
+        logger=log,
+        func_name="hook.get_metering_archive"
+    )
+
+    if archive is None:
+        archive = []
```
Comment (rtamalin) [inline csp_billing_adapter/archive.py:86]: Note that this line isn't exercised by the unit tests right now, and the use of memory_archive.copy() in memory_archive.py's get_metering_archive() implementation means you can't just initialise it to None in cba_pm() in tests/unit/conftest.py.

Not a major issue, but a little tweaking to them memory_archive.py implementation could allow 100% code coverage ;-)

**Review by rtamalin — APPROVED** [review]
LGTM

Minor suggestion in comments.

File: csp_billing_adapter/archive.py (hunk 1 above)
Comment (smarlowucf) [reply csp_billing_adapter/archive.py:86]: Expectation is that plugins return property empty value from `get_metering_archive`, i.e. `[]`. This is mainly a backup clause. I think accepting < 100% here is okay. If we modify the memory archive to return "None" by default then there's a bad example in code that might be referenced when another plugin is written. Granted having the clause makes it possible in the first place. Not sure what's best here. So I'd leave it as is.

**Review by rjschwei — APPROVED** [review]

**Review by KeithMnemonic — APPROVED** [review]
LGTM

**Pull request merged**